### PR TITLE
`<xstring>`: `basic_string::_Is_elem_cptr` should be a variable template

### LINKED ISSUE #2898: `<xstring>`: `basic_string::_Is_elem_cptr` should be a variable template
`_Is_elem_cptr<T>` is currently an alias for  `bool_constant<stuff>`. We only ever use it as `_Is_elem_cptr<T>::value`. It could more simply be a `constexpr bool` variable template initialized with `stuff` that we access directly.

https://github.com/microsoft/STL/blob/04ee87830a404437f1691d6a3e8809e4986d8493/stl/inc/xstring#L2473-L2474

### PR DESCRIPTION
**Pull request by YairBorn**
fixes #2898

### CODE REVIEW

**Comment by YairBorn** [comment]
it's related to this issue
https://github.com/microsoft/STL/issues/2898

**Comment by cpplearner** [comment]
You can see how existing variable templates are defined, e.g. : https://github.com/microsoft/STL/blob/b7e0365b5e2f9de738e8fbea42b6cc6880efe531/stl/inc/xtr1common#L189-L190

But since `_Is_elem_cptr` is a member, you should use `static` in place of `_INLINE_VAR`.

After making `_Is_elem_cptr` a variable template, you should also change all occurences of `_Is_elem_cptr<`*`something`*`>::value` to `_Is_elem_cptr<`*`something`*`>`.

**Comment by YairBorn** [comment]
> You can see how existing variable templates are defined, e.g. :
> 
> https://github.com/microsoft/STL/blob/b7e0365b5e2f9de738e8fbea42b6cc6880efe531/stl/inc/xtr1common#L189-L190
> 
> But since `_Is_elem_cptr` is a member, you should use `static` in place of `_INLINE_VAR`.
> 
> After making `_Is_elem_cptr` a variable template, you should also change all occurences of `_Is_elem_cptr<`_`something`_`>::value` to `_Is_elem_cptr<`_`something`_`>`.

i think i fixed it in https://github.com/microsoft/STL/pull/2980/commits/daddfb221aab55821c764e2de2b1d13ff3c98ae0
can you please take a look on what i did?

File: stl/inc/xstring [hunk 1]
```diff
@@ -2471,7 +2471,7 @@ private:
     static constexpr size_t _Memcpy_val_size   = sizeof(_Scary_val) - _Memcpy_val_offset;
 
     template <class _Iter>
-    using _Is_elem_cptr = bool_constant<_Is_any_of_v<_Iter, const _Elem* const, _Elem* const, const _Elem*, _Elem*>>;
+    static constexpr bool _Is_elem_cptr = bool_constant<_Is_any_of_v<_Iter, const _Elem* const, _Elem* const, const _Elem*, _Elem*>>;;
```
Comment (frederick-vs-ja) [inline stl/inc/xstring:2474]: Note that [`bool_constant` is an alias template](https://en.cppreference.com/w/cpp/types/integral_constant#Helper_alias_templates) which can't be used as an initializer. I think what we need is its argument which is a `bool` constant value.
And it's unnecessary to repeat `;`.

Comment (YairBorn) [reply stl/inc/xstring]: > Note that [`bool_constant` is an alias template](https://en.cppreference.com/w/cpp/types/integral_constant#Helper_alias_templates) which can't be used as an initializer. I think what we need is its argument which is a `bool` constant value.

Ok, i should I fix it?

Comment (cpplearner) [reply stl/inc/xstring]: Yeah, `bool_constant` was needed when `_Is_elem_cptr` was defined as an alias template. Now `_Is_elem_cptr` is a variable template, it should be removed.

Comment (YairBorn) [reply stl/inc/xstring]: >And it's unnecessary to repeat `;`.
I just fixed it.

Comment (YairBorn) [reply stl/inc/xstring]: > Yeah, `bool_constant` was needed when `_Is_elem_cptr` was defined as an alias template. Now `_Is_elem_cptr` is a variable template, `bool_constant` should be removed.

So it needs to look something like this? :
`static constexpr bool _Is_elem_cptr =_Is_any_of_v<_Iter, const _Elem* const, _Elem* const, const _Elem*, _Elem*>;`

Comment (cpplearner) [reply stl/inc/xstring]: Right, but remember to run `clang-format` to get the file correctly formatted.

Comment (frederick-vs-ja) [reply stl/inc/xstring]: > So it needs to look something like this? :
> `static constexpr bool _Is_elem_cptr =_Is_any_of_v<_Iter, const _Elem* const, _Elem* const, const _Elem*, _Elem*>;`

Yep. Just go ahead.

Comment (YairBorn) [reply stl/inc/xstring]: > Yep. Just go ahead.

Done.

**Comment by frederick-vs-ja** [comment]
Great, and further improvement can be done.
Note that
> You can (and should) link your pull request to this issue using [GitHub's close/fix/resolve syntax](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).
> (in the PR description not the commit message)

**Review by frederick-vs-ja — APPROVED** [review]

**Review by CaseyCarter — APPROVED** [review]

File: stl/inc/xstring [hunk 2]
```diff
@@ -2471,7 +2471,7 @@ private:
     static constexpr size_t _Memcpy_val_size   = sizeof(_Scary_val) - _Memcpy_val_offset;
 
     template <class _Iter>
-    using _Is_elem_cptr = bool_constant<_Is_any_of_v<_Iter, const _Elem* const, _Elem* const, const _Elem*, _Elem*>>;
+    static constexpr bool _Is_elem_cptr = _Is_any_of_v<_Iter, const _Elem* const, _Elem* const, const _Elem*, _Elem*>;
```
Comment (miscco) [inline stl/inc/xstring:2474]: Conventionally, variable templates are suffixed with `_v`

I believe we should rename this to `_Is_elem_cptr_v`

Comment (CaseyCarter) [reply stl/inc/xstring:2474]: > Conventionally, variable templates are suffixed with `_v`

Not in this library.
 
> I believe we should rename this to `_Is_elem_cptr_v`

No thank you.

Comment (CaseyCarter) [reply stl/inc/xstring:2474]: (All the other kids in the hall circle around @miscco and @CaseyCarter, as the chanting begins: "Fight, fight, fight!")

**Review by StephanTLavavej — APPROVED** [review]

**Review by strega-nil-ms — APPROVED** [review]

**Review by frederick-vs-ja — APPROVED** [review]

**Review by miscco — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Comment by StephanTLavavej** [comment]
Unfortunately, while this change is correct, it triggers a `/clr:pure` compiler bug which is unlikely to ever be fixed:

```
test.obj : fatal error LNK1179: invalid or corrupt file: duplicate COMDAT '???__E??$_Is_elem_cptr@PAD@?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@$$Q0_NB@?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@YMXXZ@?A0xbffd14b2@@$$FYMXXZ'
```

(Noticed only when mirroring to the MSVC-internal repo, as we haven't gotten around to porting the `/clr:pure` build to GitHub yet.)

**Comment by YairBorn** [comment]
> Unfortunately, while this change is correct, it triggers a `/clr:pure` compiler bug which is unlikely to ever be fixed:
> 
> ```
> test.obj : fatal error LNK1179: invalid or corrupt file: duplicate COMDAT '???__E??$_Is_elem_cptr@PAD@?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@$$Q0_NB@?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@YMXXZ@?A0xbffd14b2@@$$FYMXXZ'
> ```
> 
> (Noticed only when mirroring to the MSVC-internal repo, as we haven't gotten around to porting the `/clr:pure` build to GitHub yet.)

There's no solution to this compiler error?

**Comment by jovibor** [comment]
> while this change is correct, it triggers a `/clr:pure` compiler bug

According to the [docs](https://docs.microsoft.com/en-us/cpp/build/reference/clr-common-language-runtime-compilation?view=msvc-170):

> /clr:pure is deprecated. The option is removed in Visual Studio **2017** and later.

It was deprecated long time ago, so why even bother? 🥱

**Comment by miscco** [comment]
> Unfortunately, while this change is correct, it triggers a `/clr:pure` compiler bug which is unlikely to ever be fixed:
> 
> ```
> test.obj : fatal error LNK1179: invalid or corrupt file: duplicate COMDAT '???__E??$_Is_elem_cptr@PAD@?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@$$Q0_NB@?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@YMXXZ@?A0xbffd14b2@@$$FYMXXZ'
> ```
> 
> (Noticed only when mirroring to the MSVC-internal repo, as we haven't gotten around to porting the `/clr:pure` build to GitHub yet.)

Urgh, Now it remember that I had that change in the `ASAN string` branch and pulled it out because of this failure :(

**Comment by StephanTLavavej** [comment]
@YairBorn
> There's no solution to this compiler error?

Unfortunately, there isn't - we aren't aware of any workarounds, and the compiler team hasn't found a way to fix this longstanding issue (which involves name mangling, and hence binary compatibility). :crying_cat_face:

While we can't merge this change, this code could instead be commented as to why we aren't using a `static constexpr` data member. Would you like to change your PR to do that? We mark issues that can be addressed in the (possibly far) future as "TRANSITION" (indicating a temporary state of affairs), followed by a short explanation. Here, `// TRANSITION, /clr:pure is incompatible with templated static constexpr data members` would be a reasonable explanation. (Note that we have a 120 column limit for code format validation.)

**Comment by YairBorn** [comment]
> @YairBorn
> > There's no solution to this compiler error?
> 
> Unfortunately, there isn't - we aren't aware of any workarounds, and the compiler team hasn't found a way to fix this longstanding issue (which involves name mangling, and hence binary compatibility). :crying_cat_face:
> 
> While we can't merge this change, this code could instead be commented as to why we aren't using a `static constexpr` data member. Would you like to change your PR to do that? We mark issues that can be addressed in the (possibly far) future as "TRANSITION" (indicating a temporary state of affairs), followed by a short explanation. Here, `// TRANSITION, /clr:pure is incompatible with templated static constexpr data members` would be a reasonable explanation. (Note that we have a 120 column limit for code format validation.)
> 

If I get you right, you want me to undo all of my changes, and add the comment you mentioned?

**Comment by StephanTLavavej** [comment]
> If I get you right, you want me to undo all of my changes, and add the comment you mentioned?

Yep, that's correct.

**Comment by YairBorn** [comment]
> > If I get you right, you want me to undo all of my changes, and add the comment you mentioned?
> 
> Yep, that's correct.

Done in https://github.com/YairBorn/STL/commit/60ae0ff8983d341e9af32cbf7ccc182f33f2bd09

**Comment by YairBorn** [comment]
> > If I get you right, you want me to undo all of my changes, and add the comment you mentioned?
> 
> Yep, that's correct.

Done in https://github.com/YairBorn/STL/commit/60ae0ff8983d341e9af32cbf7ccc182f33f2bd09, but o don't know why the tests failed...

File: stl/inc/xstring [hunk 3]
```diff
@@ -2471,6 +2471,8 @@ private:
     static constexpr size_t _Memcpy_val_size   = sizeof(_Scary_val) - _Memcpy_val_offset;
 
     template <class _Iter>
+    // TRANSITION, /clr:pure is incompatible with templated static constexpr data members
+    //static constexpr bool _Is_elem_cptr =_Is_any_of_v<_Iter, const _Elem* const, _Elem* const, const _Elem*, _Elem*>;
```
Comment (fsb4000) [inline stl/inc/xstring:2475]:
```suggestion    
    // static constexpr bool _Is_elem_cptr =_Is_any_of_v<_Iter, const _Elem* const, _Elem* const, const _Elem*, _Elem*>;
```
I run locally `clang-format` on your branch
изображение

Comment (YairBorn) [reply stl/inc/xstring:2475]: Thanks, I'll fix it

**Review by fsb4000 — APPROVED** [review]

**Comment by YairBorn** [comment]
> > If I get you right, you want me to undo all of my changes, and add the comment you mentioned?
> 
> Yep, that's correct.

Done. all tests passed! 😸

**Review by cpplearner — APPROVED** [review]

**Comment by CaseyCarter** [comment]
> Urgh, Now it remember that I had that change in the `ASAN string` branch and pulled it out because of this failure :(

I _didn't_ remember, but I do recall thinking "Why haven't we done this before?" when filing the issue. Hopefully the comment will prevent us from wasting more people's time doing this again in the future.

**Review by CaseyCarter — APPROVED** [review]

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo (again) - please notify me if any further changes are pushed.

**Comment by StephanTLavavej** [comment]
Thanks for working on this change - twice - and documenting why the STL behaves like this. Also, congratulations on your first microsoft/STL commit! :tada: :heart_eyes_cat: :rocket:

This change will ship in VS 2022 17.4 Preview 3.

**Pull request merged**